src/Compilers/CSharp/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.cs

```diff
@@ -129,6 +135,113 @@ private static bool IsExcludedFromCodeCoverage(MethodSymbol method)
+                            methodToken,
+                            fileIndex,
+                            payloadSlot,
+                            methodBodyFactory.Literal(dynamicAnalysisSpans.Length)));
+            }
+            else
+            {
+                var documents = PooledHashSet<DebugSourceDocument>.GetInstance();
+                foreach (var span in dynamicAnalysisSpans)
+                {
+                    documents.Add(span.Document);
+                }
+
+                // At this point, we should have at least one document since we have already
+                // handled the case where method has no 'real' spans (and therefore no documents) above.
+                if (documents.Count == 1)
+                {
+                    var fileIndex = methodBodyFactory.SourceDocumentIndex(documents.Single());
+
+                    createPayloadStatement =
+                        methodBodyFactory.Assignment(
+                            methodBodyFactory.Local(methodPayload),
+                            methodBodyFactory.Call(
+                                null,
+                                createPayloadForMethodsSpanningSingleFile,
+                                mvid,
+                                methodToken,
+                                fileIndex,
+                                payloadSlot,
+                                methodBodyFactory.Literal(dynamicAnalysisSpans.Length)));
+                }
+                else
+                {
+                    var fileIndicesBuilder = ArrayBuilder<BoundExpression>.GetInstance(documents.Count);
+
+                    // Sort the documents so that the emit order for the indices is always deterministic.
+                    var sortedDocuments = documents.ToArray();
+                    Array.Sort(sortedDocuments, DebugSourceDocumentComparer.Instance);
+
+                    foreach (var document in sortedDocuments)
+                    {
+                        fileIndicesBuilder.Add(methodBodyFactory.SourceDocumentIndex(document));
+                    }
+
+                    var fileIndices =
+                        methodBodyFactory.Array(
+                            methodBodyFactory.SpecialType(SpecialType.System_Int32), fileIndicesBuilder.ToImmutableAndFree());
+
+                    createPayloadStatement =
+                        methodBodyFactory.Assignment(
+                            methodBodyFactory.Local(methodPayload),
+                            methodBodyFactory.Call(
+                                null,
+                                createPayloadForMethodsSpanningMultipleFiles,
+                                mvid,
+                                methodToken,
+                                fileIndices,
+                                payloadSlot,
+                                methodBodyFactory.Literal(dynamicAnalysisSpans.Length)));
+                }
+
+                documents.Free();
+            }
```

Comment:
Minor point: Each code path ends with `createPayloadStatement = ...;` with the only differences being the file indices and method. Perhaps have locals for those two values set in each code path and a single `return createPayloadStatement = ...;` at the end of the method.